pydoctor/test/test_astbuilder.py

```diff
@@ -1612,3 +1612,79 @@ def func():
     ''', systemcls=systemcls)
     outer = mod.contents['outer']
     assert not outer.contents
+
+@systemcls_param
+def test_constant_module(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.contents['LANG'], 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.contents['LANG'], 'value') is not None
+
+@systemcls_param
+def test_constant_module_with_final(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    from typing import Final
+    lang: Final = 'fr'
+    ''', systemcls=systemcls)
+    assert getattr(mod.contents['lang'], 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.contents['lang'], 'value') is not None
+
+@systemcls_param
+def test_constant_class(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.resolveName('Clazz.LANG'), 'value') is not None
+
+
+@systemcls_param
+def test_constant_init(systemcls: Type[model.System]) -> None:
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        def __init__(**args):
+            self.LANG = 'FR'
+    ''', systemcls=systemcls)
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    assert getattr(mod.resolveName('Clazz.LANG'), 'value') is not None
+
+@systemcls_param
+def test_constant_override_warns(systemcls: Type[model.System], capsys: CapSys) -> None:
+    mod = fromText('''
+    class Clazz:
+        """Class."""
+        LANG = 'EN'
+        def __init__(**args):
+            self.LANG = 'FR'
+    ''', systemcls=systemcls, modname="mod")
+    assert getattr(mod.resolveName('Clazz.LANG'), 'kind') == model.DocumentableKind.CONSTANT
+    val = getattr(mod.resolveName('Clazz.LANG'), 'value')
+    assert val is not None
+    assert ast.literal_eval(val) == 'FR'
+
+    captured = capsys.readouterr().out
+    assert "mod:4: Assignment to constant \"LANG\" overrides previous assignment.\n" in captured
```

Comment:
This is not that bad... but I am not sure if pydoctor should go this path... this looks like a pyflakes or mypy type of warning.

Reply:
I think the purpose of this warning isn't to act as a static code checker, but to inform the user that pydoctor is disregarding the other definitions, which means that if the code is not redundant, the generated docs are incomplete.

Are there valid reasons to define a constant more than once?
- maybe the value depends on the presence of certain API (OS type, Python version or optional dependency)
- maybe a base value is assigned first and later augmented to add detail

I don't know how common these situations are. But until we know, having a warning would be useful, so we can find real-world examples of duplicate definitions to consider for refining the handling.

Reply:
One word a bout this warning. It will be useful primarily when #377 is implemented. Then it would mean that the value has been overridden and the generated repr of this value is not the original. 
Right now it does not serve any particular purpose.